src/librustc_data_structures/temp_dir.rs

```diff
@@ -0,0 +1,32 @@
+use std::path::{Path, PathBuf};
+use tempfile::TempDir as TempDirReal;
+
+/// This is used to avoid TempDir being dropped on error paths unintentionally.
+#[derive(Debug)]
+pub struct TempDir {
+    path: PathBuf,
+    // This is `None` if the directory should not be deleted; used for
+    // -Csave-temps. We don't just pass around a PathBuf to keep type safety
+    // (only creating files in temporary locations).
+    dir: Option<TempDirReal>,
```

Comment:
This is a tad convoluted. Why not `ManuallyDrop<TempDirReal>` alongside a boolean flag?